remoulade/state/backends/postgres.py

```diff
@@ -0,0 +1,133 @@
+        for key in ["args", "kwargs", "options"]:
+            if key in state_dict:
+                encoded_value = encoder.encode(state_dict[key])
+                state_dict[key] = encoded_value if sys.getsizeof(encoded_value) <= max_size else None
+        return cls(**state_dict)
+
+
+class PostgresBackend(StateBackend):
+    def __init__(
+        self,
+        *,
+        namespace: str = "remoulade-state",
+        encoder: Optional[Encoder] = None,
+        client: Optional[sessionmaker] = None,
+        max_size: int = 2000000,
+        url: Optional[str] = None,
+    ):
+        self.url = url or DEFAULT_POSTGRES_URI
+        super().__init__(namespace=namespace, encoder=encoder)
+        self.client = client or sessionmaker(create_engine(self.url))
+        self.max_size = max_size
+        self.lock = threading.Lock()
+
+    def get_state(self, message_id: str):
+        with self.client.begin() as session:
+            state = session.query(StoredState).filter_by(message_id=message_id).first()
+            if state is None:
+                return None
+            return state.as_state(self.encoder)
+
+    def set_state(self, state: State, ttl=3600):
+        with self.lock:
+            with self.client.begin() as session:
+                session.merge(StoredState.from_state(state, self.max_size, self.encoder))
+
+    def get_states(
+        self,
+        *,
+        size: Optional[int] = None,
+        offset: int = 0,
+        selected_actors: Optional[List[str]] = None,
+        selected_statuses: Optional[List[str]] = None,
+        selected_ids: Optional[List[str]] = None,
+        start_datetime: Optional[datetime.datetime] = None,
+        end_datetime: Optional[datetime.datetime] = None,
+        sort_column: Optional[str] = None,
+        sort_direction: Optional[str] = None,
+        get_groups: bool = False,
+    ):
+        with self.client.begin() as session:
+            query = session.query(StoredState)
+            if selected_actors is not None:
+                query = query.filter(StoredState.actor_name.in_(selected_actors))
+            if selected_statuses is not None:
+                query = query.filter(StoredState.status.in_(selected_statuses))
+            if selected_ids is not None:
+                query = query.filter(StoredState.message_id.in_(selected_ids))
+            if start_datetime is not None:
+                query = query.filter(StoredState.enqueued_datetime >= start_datetime)
+            if end_datetime is not None:
+                query = query.filter(StoredState.enqueued_datetime <= end_datetime)
+            if get_groups:
+                query = query.filter(StoredState.group_id.is_not(None))
+            if sort_column in State._fields and sort_direction is not None:
+                query = query.order_by(text(f"{sort_column} {sort_direction}"))
+            if size is not None:
+                query = query.offset(offset).limit(size)
+
+            query = query.all()
```

Comment:
no need to put this line. If you do, it will fetch all the lines and put them in an array. And you loop on this array below.
by default query gives you a generator so your next line will work without this one and you will avoid increasing the complexity of the function